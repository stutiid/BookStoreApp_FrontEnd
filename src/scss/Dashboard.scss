.dashboard-main-div {
  width: 100%;
  height: 100%;
}
.dashboard-app-icon {
  margin-left: 100px;
  padding-right: 5px;
}
.dashboard-search-div {
  width: 30%;
  margin: auto;
  margin-top: 10px;
  margin-bottom: 10px;
  display: flex;
  padding: 10px 15px;
  background-color: white;
}
.dashboard-search-text {
  width: 100%;
}
.dashboard-input {
  width: 90%;
}
.dashboard-search-button {
  width: 10%;
  padding-top: 15px;
}
.dashboard-location-div {
  margin: 10px 12px;
  width: 99%;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
}
