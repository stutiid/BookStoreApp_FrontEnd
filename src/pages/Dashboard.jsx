import { Component, createRef } from "react";
import {
  AppBar,
  Toolbar,
  TextField,
  Button,
  List,
  ListItem,
} from "@mui/material";
import SearchIcon from "@mui/icons-material/Search";
import MapIcon from "@mui/icons-material/Map";
import "../scss/Dashboard.scss";
import Service from "../services/Service";
import DisplayLocation from "./DisplayLocation";

/**
 * @description:- functional component to render the suggestions got from the backend according to the input in the search tab
 * @param {*} props :- containing list of suggestions and a function handler passed from the parent class.
 * @returns :- List of suggesstions to be rendered on the screen
 */
const Options = (props) => {
  const items = props.suggestions.map((item) => (
    <ListItem
      key={item.id}
      onClick={(event) => {
        props.handleClick(item.name);
      }}
    >
      {item.name}
    </ListItem>
  ));
  return <List style={{ color: "black" }}>{items}</List>;
};

/**
 * @description :- the class represents a page containing the tab to search the location by name and then display it on the map.
 */
export default class Dashboard extends Component {
  constructor(props) {
    super(props);
    this.searchRef = createRef();
    this.state = {
      data: [],
      searchName: "",
      options: [],
    };
  }

  /**
   * @description:- function to set the ref value according to the argument value and then call the handleChange function to set the search input to the
   * selected value
   * @param {*} selectedValue :- value selected by user from the suggestions
   */
  handleClick = (selectedValue) => {
    this.searchRef.value = selectedValue;
    this.handleChange();
  };
  /**
   * @description:- the function handles the state change if any in the property of the component by using set state and assigning ref value to the
   * searchName state and then calling the getSuggestions function as callback.
   */
  handleChange = () => {
    this.setState(
      {
        searchName: this.searchRef.value,
      },
      () => {
        if (this.state.searchName.length >= 1) {
          this.getSuggestions();
        }
      }
    );
  };

  /**
   * @description :- to get the suggestions from the backend according to the input given by user in the search tab
   */
  getSuggestions = () => {
    (async () => {
      let response = await new Service().getLocationByName(
        "map/name_suggestions/",
        this.state.searchName,
        "suggestions"
      );
      if (response.status === 200) {
        this.setState({ options: response.data.data });
      } else alert(response);
    })();
  };

  /**
   * @description:- to set the data with the location details fetched from the backend according to the location name given by user
   */
  handleLocation = () => {
    // this.search("map/location/", this.state.searchName);
    (async () => {
      let response = await new Service().getLocationByName(
        "map/location/",
        this.state.searchName,
        "search"
      );
      if (response.status === 200) {
        this.setState({ data: response.data.data });
      } else alert(response);
    })();
  };

  /**
   * @description :- function renders the dashboard html page
   */
  render() {
    return (
      <div className="dashboard-main-div">
        <AppBar className="dashboard-app-bar" position="static" style={{ background: "#00131c" }}>
          <Toolbar>
            <MapIcon />
            <span>
              <h3>Map Location</h3>
            </span>
            <div className="dashboard-search-div">
              <div className="dashboard-input">
                <TextField
                  className="dashboard-search-text"
                  variant="standard"
                  label="Search"
                  name="searchName"
                  value={this.state.searchName}
                  inputRef={(input) => (this.searchRef = input)}
                  onChange={this.handleChange}
                  required
                />
                <Options
                  suggestions={this.state.options}
                  handleClick={this.handleClick}
                />
              </div>
              <div className="dashboard-search-button">
                <Button onClick={this.handleLocation}>
                  <SearchIcon />
                </Button>
              </div>
            </div>
          </Toolbar>
        </AppBar>
        <div className="dashboard-location-div">
          {this.state.data.map((location) => (
            <DisplayLocation data={location} key={location.id} />
          ))}
        </div>
      </div>
    );
  }
}
