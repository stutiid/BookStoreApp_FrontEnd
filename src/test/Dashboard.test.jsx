import "./setupTest";
import Dashboard from "../pages/Dashboard";
import { shallow } from "enzyme";

describe("tests", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Dashboard />);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it("should render initial layout", () => {
    expect(wrapper.getElements()).toMatchSnapshot();
  });

  it("has a app bar with menu", () => {
    expect(wrapper.find(".dashboard-app-bar")).toBeTruthy();
  });

  it("has a div containing search tab", () => {
    expect(wrapper.find(".dashboard-search-div")).toBeTruthy();
  });

  it("has a div containing search button", () => {
    expect(wrapper.find(".dashboard-search-button")).toBeTruthy();
  });

  it("contain a child component to display the map with selected location", () => {
    expect(wrapper.find(".dashboard-location-div")).toBeTruthy();
  });

  it("contain a div to represent the sugesstions", () => {
    expect(wrapper.find(".dashboard-input")).toBeTruthy();
  });
});
