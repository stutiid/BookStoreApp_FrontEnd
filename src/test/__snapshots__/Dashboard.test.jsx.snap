// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests should render initial layout 1`] = `
Array [
  <div
    className="dashboard-main-div"
  >
    <ForwardRef(AppBar)
      className="dashboard-app-bar"
      position="static"
      style={
        Object {
          "background": "#00131c",
        }
      }
    >
      <ForwardRef(Toolbar)>
        <Memo />
        <span>
          <h3>
            Map Location
          </h3>
        </span>
        <div
          className="dashboard-search-div"
        >
          <div
            className="dashboard-input"
          >
            <ForwardRef(TextField)
              className="dashboard-search-text"
              inputRef={[Function]}
              label="Search"
              name="searchName"
              onChange={[Function]}
              required={true}
              value=""
              variant="standard"
            />
            <Options
              handleClick={[Function]}
              suggestions={Array []}
            />
          </div>
          <div
            className="dashboard-search-button"
          >
            <ForwardRef(Button)
              onClick={[Function]}
            >
              <Memo />
            </ForwardRef(Button)>
          </div>
        </div>
      </ForwardRef(Toolbar)>
    </ForwardRef(AppBar)>
    <div
      className="dashboard-location-div"
    />
  </div>,
]
`;
